[Wallet] Fix staking balance calculation

random-zebra opened the pull request:
exclude delegated balance and locked coins (if any)

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1731,7 +1731,8 @@ CAmount CWallet::GetColdStakingBalance() const
 {
-    return GetBalance() + (fIncludeColdStaking ? GetColdStakingBalance() : 0);
+    const CAmount& nBalance = GetBalance() - GetDelegatedBalance() - GetLockedCoins();
+    return nBalance + (fIncludeColdStaking ? GetColdStakingBalance() : 0);
```
Each of the `GetBalanceXXX` methods call are going through all of the txs inside the wallet.

Would be better to have a single method filtering the transactions that needs to be filter only for this (one single loop over all of the txs and not three).

random-zebra replied:
absolutely, nice idea.

random-zebra commented:
Refactored `GetStakingBalance` using a single loop as per @furszy 's suggestion.
Also fixed two bugs:
- do not count those coins that don't have the required stake depth
- prevent possible negative values for balances after the loop

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1710,7 +1710,8 @@ CAmount CWallet::loopTxsBalance(std::function<void(const uint256&, const CWallet
     }
-    return nTotal;
+    // Amounts can be negative (method could subtract them), but final balance shouldn't.
+    return std::max((CAmount) 0, nTotal);
```
Hmm, I'm not really in favor of this. It could end up hiding other issues in the future. `loopTxsBalance` method should be an abstract loop over the txs only, abstracted from what could happen inside the method.

 Better to have it only in the method that needs it.

random-zebra replied:
Balances should never be negative, but I can see the point. Moved.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -1731,7 +1732,16 @@ CAmount CWallet::GetColdStakingBalance() const
 CAmount CWallet::GetStakingBalance(const bool fIncludeColdStaking) const
 {
-    return GetBalance() + (fIncludeColdStaking ? GetColdStakingBalance() : 0);
+    AssertLockHeld(cs_main);
```
This shouldn't be here, it should be the opposite `AssertNotLockHeld`.

The `cs_main` lock is being acquired inside the `loopTxsBalance`.

random-zebra replied:
Right. Didn't notice `loopTxsBalance` already acquired the lock.
Fixed

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -2035,7 +2045,11 @@ bool less_then_denom(const COutput& out1, const COutput& out2)
+    CAmount nBalance = 0;
+    {
+        LOCK(cs_main);
+        nBalance = GetStakingBalance(GetBoolArg("-coldstaking", true));
```
No need to acquire the LOCK here, it's acquired inside `loopTxsBalance`.

furszy commented on src/wallet/wallet.cpp:
```diff
@@ -3663,6 +3677,7 @@ int CMerkleTx::SetMerkleBranch(const CBlock& block)
 
 int CMerkleTx::GetDepthInMainChain(bool enableIX) const
 {
+    AssertLockHeld(cs_main);
```
I don't see any value on this AssertLockHeld, just a duplicated check. it's already being checked inside the `GetDepthInMainChain(pindexRet, enableIX)` that it's two lines below.

random-zebra commented:
Fixed LOCKs.

Note: with #1373 `GetStakingBalance` becomes useless.
We could keep it anyway to call it on demand by `getstakingstatus` with #1368.
